connections/README.md

```diff
@@ -263,6 +267,49 @@ indicating whether the handler supports the protocol. This allows more flexible
 behavior than exact literal matching, which is the default behavior if no match
 function is provided.
 
+## Stream Behaviour
+
+This section specifies the expected behaviour of a libp2p stream.
+
+### States
+
+A libp2p stream can be in one of four states:
+
+- Open for reading and writing
+- Open for reading only
+- Open for writing only
+- Closed
+
+The default state after opening a new stream is "Open for reading and writing".
+Either party can at any point close the reading or writing side.
+
+The "Closed" state is the product of both halfs (read and write) of the stream being closed.
+This can happen because one party closes both halfs but also as a combination
+of both parties closing their write-half (which automatically closes other's read-half).
+
+Writing to a write-closed stream SHOULD fail with an error clearly signalling the closed state.
+
+Reading from a read-closed stream SHOULD succeed by reading 0 bytes.
```

Comment:
Intuitively this is to implementation specific. E.g. some may express this through a special event, some might express this through `0 bytes`. I don't think we should dictate the way of how an implementation expresses this.

Reply:
```suggestion
Reading from a read-closed stream SHOULD succeed.
```

Would that be better?